src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/SelectBranchTrait.java

```diff
@@ -0,0 +1,53 @@
+package com.atlassian.bitbucket.jenkins.internal.scm;
+
+import com.atlassian.bitbucket.jenkins.internal.trigger.register.PullRequestStore;
+import hudson.Extension;
+import jenkins.scm.api.SCMHead;
+import jenkins.scm.api.SCMSource;
+import jenkins.scm.api.trait.SCMHeadPrefilter;
+import jenkins.scm.api.trait.SCMSourceContext;
+import jenkins.scm.api.trait.SCMSourceTrait;
+import jenkins.scm.api.trait.SCMSourceTraitDescriptor;
+import org.kohsuke.stapler.DataBoundConstructor;
+
+import javax.inject.Inject;
+
+public class SelectBranchTrait extends SCMSourceTrait {
+
+    @DataBoundConstructor
+    public SelectBranchTrait() {
+    }
+
+    @Override
+    protected void decorateContext(SCMSourceContext<?, ?> context) {
+        context.withPrefilter(new SCMHeadPrefilter() {
+            @Override
+            public boolean isExcluded(SCMSource scmSource,
+                                      SCMHead scmHead) {
+
+                if (scmSource instanceof CustomGitSCMSource) {
+                    BitbucketSCMRepository bbsRepository =
+                            ((CustomGitSCMSource) scmSource).getRepository();
+                    PullRequestStore pullRequestStore =
+                            ((DescriptorImpl) getDescriptor()).pullRequestStore;
+                    //TODO: add logging
+                    return !pullRequestStore.hasOpenPullRequests(scmHead.getName(), bbsRepository);
+                }
+                return true;
+            }
+        });
+    }
+
+    @Extension
+    public static class DescriptorImpl extends SCMSourceTraitDescriptor {
+
+        @Inject
+        private PullRequestStore pullRequestStore;
+
+        @Override
+        public String getDisplayName() {
+            return "Only build branches with open pull requests";
```

Comment:
Small note, but I think we need to consider the copy here. This text makes the behaviour look generic, as though it may apply to all SCMs when it is specifically working with Bitbucket Server- so I think this text should reflect that.

Reply:
I agree, I think the way forward is that we merge this to the feature branch, and once that branch is ready to go to master, we sit down and go over the UI copy (and elements etc) and perhaps also seek input from Cameron.